src/encoding/byte_stream_split/decoder.rs

```diff
@@ -0,0 +1,61 @@
+use std::marker::PhantomData;
+use crate::error::Error;
+use crate::types::NativeType;
+
+/// Decodes according to [Byte Stream Split](https://github.com/apache/parquet-format/blob/master/Encodings.md#byte-stream-split-byte_stream_split--9).
+/// # Implementation
+/// This struct does not allocate on the heap.
+#[derive(Debug)]
+pub struct Decoder<'a, T: NativeType> {
+    values: &'a [u8],
+    num_elements: usize,
+    current: usize,
+    element_size: usize,
+    element_type: PhantomData<T>
+}
+
+impl<'a, T: NativeType> Decoder<'a, T> {
+    pub fn try_new(values: &'a [u8]) -> Result<Self, Error> {
+        let element_size = std::mem::size_of::<T>();
+        let values_size = values.len();
+        if values_size % element_size != 0 {
+            return Err(Error::oos("Value array is not a multiple of element size"));
+        }
+        let num_elements = values.len() / element_size;
+        Ok(Self {
+            values,
+            num_elements,
+            current: 0,
+            element_size,
+            element_type: PhantomData
+        })
+    }
+}
+
+impl<'a, T: NativeType> Iterator for Decoder<'a, T> {
+    type Item = Result<T, Error>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current >= self.num_elements {
+            return None
+        }
+
+        let mut buffer = vec![0_u8; self.element_size];
```

Comment:
isn't this an allocation on the heap? (the "This struct does not allocate on the heap." on the discussion) Would it make sense allocate this once, and re-use it so that we do not allocate on every call of `next`?

Reply:
I'm not familiar with how rust works, but I hoped because the lifetime of this slice is within the function scope hence it doesn't need to escape to the healp and would be allocated on the stack. To be honest I wouldn't know how to check this.

Ideally I would have liked to allocate a fixed size slice on Decoder itself to reuse, but seems that it's not possible do so something like:

```
struct Foo<T> {
   buf: [u8; std::mem::size_of::<T>()]
}
```

I guess I'll do what you are suggesting and allocate a slice in try_new.